### PR TITLE
Hid update protocol

### PR DESCRIPTION
**Pull request by sctanf**
switch to 16 byte packets
fill in more data for hid devices and hid trackers
change unique identifier for hid trackers
add support for exponential map quantized quaternions
http://marc-b-reynolds.github.io/quaternions/2017/05/02/QuatQuantPart1.html#fnref:pos:3
https://github.com/Marc-B-Reynolds/Stand-alone-junk/blob/559bd78893a3a95cdee1845834c632141b945a45/src/Posts/quatquant0.c#L898

### CODE REVIEW

File: server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt [hunk 1]
```diff
@@ -56,7 +56,7 @@ class Tracker @JvmOverloads constructor(
 	 * Has solved position + rotation (Vive trackers)
 	 */
 	val isComputed: Boolean = false,
-	val imuType: IMUType? = null,
+	var imuType: IMUType? = null, // TrackersHID has to assign after the device is created
```
Comment (ImUrX) [inline server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt:59]: but a tracker should be initialized when the tracker info is received, not when the device is created...

File: server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt [hunk 2]
```diff
@@ -199,64 +212,197 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 				if (dataReceived.isNotEmpty()) {
 					// Process data
 					// TODO: make this less bad
-					if (dataReceived.size % 20 != 0) {
+					if (dataReceived.size % 16 != 0) { // TODO: packet size 20 is valid (old)
```
Comment (ImUrX) [inline server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt:215]: use a `const` for this `PACKET_SIZE` please

**Review by ImUrX — CHANGES_REQUESTED** [review]

File: server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt (hunk 1 above)
Comment (sctanf) [reply server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt:59]: i can wait for the info packet before creating the tracker, but that depends if the tracker has time to send the info packet, it must be awake but not moving

Comment (ImUrX) [reply server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt:59]: this sounds like something that should be fixed on esb protocol.
idk, i would like more opinions if we should let trackers be lazily updated or for a tracker to be initialized it should contain the info for its initialization

Comment (sctanf) [reply server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt:59]: i can update it to wait for the info packet, but it means the tracker may not get picked up quickly
a user probably does not constantly move around right after a device is started, so it is ok
devices were allowed to override its data, like MCUType, only for trackers it cannot be overridden without modifying base tracker

**Comment by sctanf** [comment]
some latency issue was appearing on receiver reconnect, maybe it is some buffer, so switched to using readAll

File: server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt (hunk 1 above)
Comment (sctanf) [reply server/core/src/main/java/dev/slimevr/tracking/trackers/Tracker.kt]: after some more use, it doesnt seem like it is much of an issue as i thought

**Review by Eirenliel — APPROVED** [review]

**Pull request merged**